src/storage/scripting.cc

```diff
@@ -1470,8 +1488,73 @@ Status CreateFunction(Server *srv, const std::string &body, std::string *sha, lu
   }
   lua_setglobal(lua, funcname);
 
+  // Cache the flags of the current script
+  uint64_t script_flags = 0;
+  if (auto pos = body.find('\n'); pos != std::string::npos) {
+    auto first_line = body.substr(0, pos);
+    if (util::HasPrefix(first_line, "#!lua")) {
+      uint64_t shebang_flags = 0;
+      if (auto s = ExtractFlagsFromShebang(first_line, shebang_flags); !s.IsOK()) {
+        lua_pop(lua, 1); /* remove the error handler from the stack. */
+        return s;
+      }
+      script_flags |= shebang_flags;
+    } else {
+      // scripts without #! can run commands that access keys belonging to different cluster hash slots,
+      // but ones with #! inherit the default flags, so they cannot.
+      script_flags |= ScriptFlags::kScriptAllowCrossSlotKeys;
+    }
+  }
+  srv->CacheScriptFlags(*sha, script_flags);
```

Comment:
why is it stored in server instead of lua global variables?

we already have a cache mechanism for `CreateFunction` [here](https://github.com/apache/kvrocks/blob/236ec7a3257a5b03d6006e06b672f64f5337bdcf/src/storage/scripting.cc#L620), i.e. if the script is loaded in the lua vm (recall that we have multiple vms) `CreateFunction` will not be called.

Reply:
I think both are possible. If it is stored in server, different Lua vm can share it. There is no need to create a global variable for each vm.

> if the script is loaded in the lua vm (recall that we have multiple vms) CreateFunction will not be called.

I don’t quite understand this. If the script has been loaded, then CreateFunction has been called for the first time and the flag has also been stored. 

Can you give a more detailed example or point out the key code directly? Maybe I've overlooked something.

Reply:
I think we can try to avoid adding more members to class Server if possible. (it's better for maintanace that it can only be accessed in the lua vm rather than almost all the codebase.)

Also in lua vm we can ensure it's accessed from just one thread (since it's not thread-safe), and it will be deleted after the lua vm is reset.

Reply:
Okay, that sounds good, I'll try to redesign it